src/test/java/com/xwiki/authentication/saml/XWikiGroupManagerTest.java

```diff
@@ -0,0 +1,164 @@
+package com.xwiki.authentication.saml;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.classes.BaseClass;
+import com.xpn.xwiki.web.XWikiRequest;
+import com.xpn.xwiki.web.XWikiResponse;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.slf4j.LoggerFactory;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+
+import javax.servlet.http.HttpSession;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class XWikiGroupManagerTest {
+    @Test
+    public void addUserToGroup_ShouldAddUserToGivenGroup() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+        .whenAddedToGroup()
+        .then()
+            .user("ArthurDent")
+            .isInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void removeUserFromGroup_ShouldRemoveUserFromGivenGroup() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+            .userIsAlreadyInGroup("ArthurDent","StarshipTroopers")
+        .whenRemoveFromGroup()
+        .then()
+            .user("ArthurDent")
+            .isntInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void userAlreadyMemberOfGroup_ShouldIgnoreTheNewMembership() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+            .userIsAlreadyInGroup("ArthurDent","StarshipTroopers")
+        .whenAddedToGroup()
+        .then()
+            .user("ArthurDent")
+            .isInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void removeUserFromGroupWhereHeIsnt_ShouldNotChangeGroupMembership() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+        .whenRemoveFromGroup()
+        .then()
+            .user("ArthurDent")
+            .isntInGroup("StarshipTroopers");
+    }
```

Comment:
All of these tests have the same problem. The thing you are testing is happening on the `given` and the when is implicitly referencing it (read: inside your DSL, where the test writer should not touch), making the test harder to read. Refactor this so these happen on the `when`.

Reply:
The action being tested happens on when* methods. Anyone trying to understand the code, will browse the when method and will find the action right there. So, from my point of view, it is pretty is to understand.